platform/mellanox/mlnx-sfpd/scripts/mlnx-sfpd

```diff
@@ -101,34 +100,65 @@ class MlnxSfpd:
-        self.user_channel_p.channel.fd = self.rx_fd_p
+            if not sx_api_open_pass:
+                rc, self.handle = sx_api_open(None)
+                if rc == SX_STATUS_SUCCESS:
+                    sx_api_open_pass = True
+                    retry = 0
+                else:    
+                    log_warning("failed to open SDK API handle with err code {}... retrying {}".format(rc,retry))
+                    retry += 1
+                    time.sleep(2 ** retry)
+                    if retry > self.SX_OPEN_RETRIES:
+                        raise RuntimeError("failed to open SDK API handle after {} retries".format(retry))
+                    else:
+                        continue
+        
+            if not sx_api_host_ifc_open_pass:
+                rc = sx_api_host_ifc_open(self.handle, self.rx_fd_p)
+                if rc == SX_STATUS_SUCCESS:
+                    self.user_channel_p.type = SX_USER_CHANNEL_TYPE_FD
+                    self.user_channel_p.channel.fd = self.rx_fd_p
+                    sx_api_host_ifc_open_pass = True
+                    retry = 0
+                else:
+                    log_warning("sx_api_host_ifc_open failed with error code {}... retrying {}".format(rc,retry))
+                    retry += 1
+                    time.sleep(2 ** retry)
+                    if retry > self.SX_OPEN_RETRIES:
+                        sx_api_close(self.handle)
+                        raise RuntimeError("sx_api_host_ifc_open failed with error after {} retries".format(retry))
+                    else:
+                        continue
 
-        rc = sx_api_host_ifc_trap_id_register_set(self.handle,
-                                                  SX_ACCESS_CMD_REGISTER,
-                                                  self.swid,
-                                                  SX_TRAP_ID_PMPE,
-                                                  self.user_channel_p)
-        if rc != SX_STATUS_SUCCESS:
-            raise RuntimeError("sx_api_host_ifc_trap_id_register_set exited with error, rc {}".format(c))
+            if not sx_api_host_ifc_trap_id_register_set_pass:
+                rc = sx_api_host_ifc_trap_id_register_set(self.handle,
+                                                          SX_ACCESS_CMD_REGISTER,
+                                                          self.swid,
+                                                          SX_TRAP_ID_PMPE,
+                                                          self.user_channel_p)
+                if rc == SX_STATUS_SUCCESS:
+                    sx_api_host_ifc_trap_id_register_set_pass = True
+                    break
+                else:
+                    log_warning("sx_api_host_ifc_trap_id_register_set failed with error code {}... retrying {}".format(rc,retry))
+                    retry += 1
+                    time.sleep(2 ** retry)
+                    if retry > self.SX_OPEN_RETRIES:
```

Comment:
`SX_OPEN_RETRIES` is set to 20. Will it be too long to wait for 2 ^ 20 seconds?

Reply:
Echo Shuotian. It doesn't make sense to sleep 12.1 days and retry. One suggestion is keep (2^retry) until a maximal value.

---
In reply to: [306872036](https://github.com/Azure/sonic-buildimage/pull/3193#discussion_r306872036) [](ancestors = 306872036)

Reply:
set max retry time to 10, max waiting time for each API will be 1024 sec